pgbouncer/kubedb_client_builder.go

```diff
@@ -0,0 +1,175 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package pgbouncer
+
+import (
+	"context"
+	"database/sql"
+	"errors"
+	"fmt"
+
+	api "kubedb.dev/apimachinery/apis/kubedb/v1alpha2"
+
+	"log"
+	"time"
+
+	_ "github.com/lib/pq"
+	core "k8s.io/api/core/v1"
+	"k8s.io/klog/v2"
+	"kmodules.xyz/client-go/tools/certholder"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"xorm.io/xorm"
+)
+
+const (
+	DefaultPgBouncerDB = "pgbouncer"
+)
+
+type KubeDBClientBuilder struct {
+	kc          client.Client
+	db          *api.PgBouncer
+	url         string
+	podName     string
+	pgBouncerDB string
+}
+
+func NewKubeDBClientBuilder(kc client.Client, db *api.PgBouncer) *KubeDBClientBuilder {
+	return &KubeDBClientBuilder{
+		kc: kc,
+		db: db,
+	}
+}
+
+func (o *KubeDBClientBuilder) WithPod(podName string) *KubeDBClientBuilder {
+	o.podName = podName
+	return o
+}
+
+func (o *KubeDBClientBuilder) WithURL(url string) *KubeDBClientBuilder {
+	o.url = url
+	return o
+}
+
+func (o *KubeDBClientBuilder) WithPgBouncerDB(pgDB string) *KubeDBClientBuilder {
+	o.pgBouncerDB = pgDB
+	return o
+}
+
+func (o *KubeDBClientBuilder) GetPgbouncerXormClient() (*XormClient, error) {
+	connector, err := o.getConnectionString()
+	if err != nil {
+		return nil, err
+	}
+
+	engine, err := xorm.NewEngine("postgres", connector)
+	if err != nil {
+		return nil, fmt.Errorf("failed to generate pgbouncer client using connection string: %v", err)
+	}
+	_, err = engine.Query("SHOW HELP")
+	if err != nil {
+		return nil, fmt.Errorf("failed to run query: %v", err)
+	}
+	return &XormClient{engine}, nil
+}
+
+func (o *KubeDBClientBuilder) getURL() string {
+	return fmt.Sprintf("%s.%s.%s.svc", o.podName, o.db.GoverningServiceName(), o.db.Namespace)
+}
+
+func (o *KubeDBClientBuilder) getPgBouncerAuthCredentials() (string, string, error) {
+	if o.db.Spec.AuthSecret == nil {
+		return "", "", errors.New("no database secret")
+	}
+	var secret core.Secret
+	err := o.kc.Get(context.TODO(), client.ObjectKey{Namespace: o.db.Namespace, Name: o.db.Spec.AuthSecret.Name}, &secret)
+	if err != nil {
+		return "", "", err
+	}
+	return string(secret.Data[core.BasicAuthUsernameKey]), string(secret.Data[core.BasicAuthPasswordKey]), nil
+}
+
+func (o *KubeDBClientBuilder) GetPgBouncerClient() (*Client, error) {
+	connector, err := o.getConnectionString()
+	if err != nil {
+		return nil, err
+	}
+	// connect to database
+	db, err := sql.Open("postgres", connector)
+	if err != nil {
+		return nil, err
+	}
+
+	// ping to database to check the connection
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second)
+	defer cancel()
+
+	if _, err := db.QueryContext(ctx, "SHOW HELP;"); err != nil {
+		klog.Infoln(db)
+		klog.Infoln("here is error", connector)
```

Comment:
can we improve error msg? better to use klog.Errorf() or klog.Error()

Reply:
It's for dbg, forgot to remove